sync/class.jetpack-sync-full.php

```diff
@@ -163,27 +163,47 @@ private function enqueue_all_terms() {
 	}
 
 	private function enqueue_all_posts() {
+		global $wpdb;
 		$this->set_status( 'posts', 0 );
+		$post_type_sql = Jetpack_Sync_Defaults::get_blacklisted_post_types_sql();
+		$this->enqueue_all_ids_as_action( 'jetpack_full_sync_posts', $wpdb->posts, 'ID', $post_type_sql, 'posts' );
+		$this->set_status( 'posts', 100 );
+	}
+
+	private function enqueue_all_ids_as_action( $action_name, $table_name, $id_field, $where_sql, $status_name ) {
 		global $wpdb;
 
-		// I hope this is never bigger than RAM...
-		$post_type_sql = Jetpack_Sync_Defaults::get_blacklisted_post_types_sql();
-		$post_ids      = $wpdb->get_col( "SELECT id FROM $wpdb->posts WHERE $post_type_sql" ); // Should we set a limit here?
+		if ( ! $where_sql ) {
+			$where_sql = "1 = 1";
+		}
 
-		// Request posts in groups of N for efficiency
-		$chunked_post_ids = array_chunk( $post_ids, self::ARRAY_CHUNK_SIZE );
+		$total = $wpdb->get_var( "SELECT count(*) FROM {$table_name} WHERE {$where_sql}" );
+		$items_per_page = 500;
+		$page = 1;
+		$counter = 1;
+		$offset = ( $page * $items_per_page ) - $items_per_page;
 
-		$counter = 0;
-		$total   = count( $chunked_post_ids );
+		while( $ids = $wpdb->get_col( "SELECT {$id_field} FROM {$table_name} WHERE {$where_sql} ORDER BY {$id_field} asc LIMIT {$offset}, {$items_per_page}" ) ) {
+			print_r($ids, true);
+			// Request posts in groups of N for efficiency
+			$chunked_ids = array_chunk( $ids, self::ARRAY_CHUNK_SIZE );
 
-		// Send each chunk as an array of objects
-		foreach ( $chunked_post_ids as $chunk ) {
-			$this->set_status( 'posts', ( $counter / $total ) * 100 );
-			do_action( 'jetpack_full_sync_posts', $chunk );
-			$counter += 1;
-		}
+			// Send each chunk as an array of objects
+			foreach ( $chunked_ids as $chunk ) {
+				$this->set_status( $status_name, ( $counter / $total ) * 100 );
+				/**
+			 	 * Fires with a chunk of object IDs during full sync.
+			 	 * These are expanded to full objects before upload
+			 	 *
+			 	 * @since 4.1
```

Comment:
Answered myself, should be 4.1.0. I'll push a PR for all of them in the codebase.